### DISCUSSION TITLE
Fullscreen mode switch

### DISCUSSION THREAD

**Discussion by Branikolog** [ihhub/fheroes2#2433] [Ideas]
Here we can discuss how can we implement a special in game switch for fullscreen mode.

**Comment by Branikolog** [comment]
As we don't have an artist, we cannot draw something new. So we need to use some existing UI solutions.
I prefer have this switch at the same place, where resolution selection happens.
Current window is not bad, but we cannot expand it to fit new elements...
We can use the same window, but change the lowest line with resolution, which duplicates the resolution from the full list above. Resolution in list is already marked by a yellow color, so we don't need a copy in bottom. 
![image](https://user-images.githubusercontent.com/55348946/102647206-70d6be00-4176-11eb-9872-ccf9882f66e6.png)
I suggest type there "Full screen" or "Window mode". If we cannot create pull-down menu there, we can simply show a small pop up window with a choice like "Choose window or full screen mode". We also can place a tip there, that player can change mode in game with the help of F4 buton.

**Comment by LeHerosInconnu** [comment]
@Branikolog.

The "Select Game Resolution:" window is accessible only when the program has just been launched and this window is not accessible directly during the game (except to save the current game and return to the home screen).

The place for this action would be more appropriate in the "System Options" window:

![System Options fullscreen mode switch fh2 01](https://user-images.githubusercontent.com/43583994/102654700-4c2c1780-4171-11eb-8bcb-c18839960193.png)

In this window the unused button can be used to choose between full-screen and windowed mode, and even to change the resolution during the game:
- left click, full screen mode, another left click windowed mode, etc., the mode is activated as soon as the user clicks the button and if it suits his liking, he clicks the "OKAY" button to exit the window.
- right click to bring up the "Select Game Resolution:" window where the user can change the resolution during the game.

All there is to do is to find a small (even temporary) graphic to put on the button. :)

**Comment by Branikolog** [comment]
@LeHerosInconnu 
I like your idea. Actually, I wanted to save that spare and useful slot in options menu. :) For something in future. :) 
Totally agreed having more adjustable options in game is nice thing. On the other hand, lots of games (if I'm not mistaken, even H5) don't support resolution switching during gameplay (resolution drop-down menu is inactive), only in Main Menu... So I can't tell it must be there.
Talking about window mode: I can't imagine, that players would use that menu each time to switch mode. We have a special keyboard button, which is more easy to use. :) 
I like current implementation of resolution selection in game. I got used to it after two launches of the game. :) We have a tip on starting application, and I always know, where to find resolution list. (which I set only once, and then continue playing with this resolution each time). I think players would also be glad to see there mode switch button, and read the tip how to use it with keyboard.
The most common pattern of switching between modes - is when player needs to check fast some other application in OS, and when we're talking about "fast" it's not like clicking a small button of GUI among other 7 tiny boxes, and switching there mode. :D
So your suggestion is nice and more GUI-friendly (looks more natural with interface of the game), but I think, we can add there a more useful option in future. Also, changing resolution with clicking over a single button is not comfortable, as on some devices the list is too huge, and you have to click till the needed resolution for eternity. :) And pray not to miss the one. :D

**Reply by LeHerosInconnu** [reply]
@Branikolog.

I see your points. :)

If you want something very fast and available everywhere without using the "F4" key, I only see the mouse, and moreover the user already has his hand on it. :)

When I was thinking about how to quickly join troops of the same type or to separate troops simply and easily in the hero's window, I thought about the "mouse gesture".

Basically, drawing signs directly with the mouse to perform predetermined actions (it can also work with the finger for touchscreens), but I thought it might be a bit "too much". :D

How do you think about that?

**Reply by Branikolog** [reply]
Drawing signs with mouse is to complicated and up-to-date to such game. :)

**Comment by LeHerosInconnu** [comment]
In the original game, in windowed mode, there is a menu that allows the user to switch to full screen mode and change the game resolution.

![Full screen menu H2 01](https://user-images.githubusercontent.com/43583994/102683008-fafe4100-41cd-11eb-9e43-4f01743829ae.png)

It is even possible to have a custom window size by moving the edges of the Windows window with the mouse.

![Full screen menu H2 02](https://user-images.githubusercontent.com/43583994/102683012-005b8b80-41ce-11eb-9253-66132529234a.png)

This menu is not available in full-screen mode, but it would be possible to make a menu (or a window, etc.) appear even in full-screen mode by a mouse action, for example by right-clicking and holding anywhere on the screen and moving the mouse to the top edge of the screen (or even any edge of the screen).

**Reply by ihhub** [reply]
Sadly SDL doesn't support menu.

**Comment by Branikolog** [comment]
@LeHerosInconnu.
I cannot say, that that your first suggestion bad. It's OK. But I'm pretty sure, I won't use it there. :) Too many actions. 
You know, I don't think, that this switch is a "musthave". I've easily survived without it. 
I think, having such switch is something like mannerliness from a developers. Game should have some options, even if people hardly ever use them. :) 
So, I prefer having it somewhere in GUI, so, I can tell, that: "We have that." The only thing I wish is that its place is not hard-to-reach, and people would definitely find it. 
I remember, while playing H3 long time ago I've accidently pressed button and turned window mode on. And I had lots of panic, while trying to find out how to make everything back.))) 
So, my idea is to place this switch to the place, player definitely know about. I wish it only look really neat in GUI.

**Comment by LeHerosInconnu** [comment]
@Branikolog.

After all you've written, I don't see much point in adding this feature.

- First, it already exists: the "F4" key.
- Second, if it's just to say: "It exists in the program." (other than with the "F4" key, and only with the use of the mouse); but it's not ergonomic, it's just crap.
- Thirdly, why would you want to put it in a window that has nothing to do with it, the "Select Game Resolution:" window; it doesn't make sense.

**Comment by Branikolog** [comment]
F4 is the best and the most comfortable solution for this feature. I believe, that implementing in ANY place of GUI of this feature won't be so comfortable, as F4. So this would be useful only for "emergency" switching or just for fun. :) (In a case we're talking about switching in GUI using mouse) 
I think, this feature should be noticeable at start of application, so user would make a decision, what mode he prefers and continue playing, holding in mind, that he can switch it with F4 or with a button "placed somewhere in GUI". :)
So I wish this switch be eye-friendly and be placed in such way, player won't forget where.
I've decided to place it in "Select game resolution" menu, as this two options affect video display of the game.
Anyway, it not necessary should be there. I can even accept this feature, for example, in CONFIG menu. But I like worse this idea, as I don't like that mess in CONFIG menu.
Adding this feature to a system option menu in game is nice idea, but I'm too greedy, to use that spare slot for such feature.))

**Comment by ihhub** [comment]
Do we agree that we put this option in system options dialog?

**Comment by ihhub** [comment]
Also I think we need to change the first message to add F4 information there.

**Comment by ihhub** [comment]
@Branikolog in this case where do we put the option? In the same window with resolutions?

**Comment by LeHerosInconnu** [comment]
@ihhub and @Branikolog.

The "Select Game Resolution:" window is simple and clear.
Why would you want to confuse the user with an action that is not really related?

Just add a sentence to the "Please remember" window to tell the user that he can use the "F4" key on the keyboard to switch from windowed mode to full screen mode:

"You can also switch from windowed mode to full screen mode and vice versa by using the "F4" key on the keyboard."

![Please remember window fh2 01](https://user-images.githubusercontent.com/43583994/102689884-e9cf2780-4201-11eb-8fc5-ed88812ead99.png)

**Reply by Branikolog** [reply]
> Why would you want to confuse the user with an action that is not really related?

Cannot say it's not related. :) 
![Display](https://user-images.githubusercontent.com/55348946/102690221-fc069100-4214-11eb-9233-e4796bdd3b42.jpg)

**Reply by LeHerosInconnu** [reply]
@Branikolog.

So this is your goal. :D

As you can see, the "Full Screen" option is not mixed with the different screen resolutions. :)

If you want something similar, the window should be renamed "Video Settings", with one part for game resolution and another separate part for full screen/windowed mode.

**Reply by Branikolog** [reply]
So in my proposal they are not mixed either.
![102647206-70d6be00-4176-11eb-9872-ccf9882f66e6(1)](https://user-images.githubusercontent.com/55348946/102691973-6291ac00-4221-11eb-9ba7-ed39853947d2.png)
This section is separated from resolutions. 
And yes, I'm dreaming of a separate "video options" menu for our project. :D

**Reply by LeHerosInconnu** [reply]
@Branikolog.

> So in my proposal they are not mixed either.

This will make this window different in its use of the "File to Load:", "File to Save:" and "Select Scenario" windows.
I'm not for that.

> This section is separated from resolutions.

It will just look like a "do-it-yourself", the window needs to be reworked properly, there's no point in doing it halfway. :)

**Reply by Branikolog** [reply]
Well, I think, for a temporary solution it's not so bad... I hope, it doesn't go far from original GUI...
![mode1](https://user-images.githubusercontent.com/55348946/102692778-acc95c00-4226-11eb-90aa-fbc66780b2ad.jpg)
![mode2](https://user-images.githubusercontent.com/55348946/102692780-adfa8900-4226-11eb-885c-524f6980c86e.jpg)

**Reply by LeHerosInconnu** [reply]
@Branikolog.

> Well, I think, for a temporary solution it's not so bad... I hope, it doesn't go far from original GUI...

If it's only temporary, why not, but I think it really doesn't fit. :face_with_head_bandage: 

What can I say, other than, the text is not centered either vertically or horizontally in the frame. :D

**Edit.**

Also, it should be "Full Screen Mode" instead of "Fullscreen Mode". :D:D

**Comment by Branikolog** [comment]
Current tip on application start is temporary. I'm not sure we will show it after 1.0 release or, let's say 1.1. I think in future we can implement display options in more noticeable way, than that door, which i found out is clickable a few months ago. :) 
But for now we can make such messages on start to make people remember the features. 
We can add a sentence to an existing window. Or even can make another window appearing after closing the one about resolutions. 
(Note that phrase "Enjoy..." should be in the last appeared window then.)

**Reply by LeHerosInconnu** [reply]
@Branikolog.

> Or even can make another window appearing after closing the one about resolutions.

Please, do not force the player to click on a dozen windows when the game opens.
A single window is more than enough.

> (Note that phrase "Enjoy..." should be in the last appeared window then.)

Yes, the sentence "Enjoy..." should be at the end of the text in the one and only window.
And I would even say with a line of space before it. :)

**Reply by Branikolog** [reply]
> Please, do not force the player to click on a dozen windows when the game opens.

There will be just two.))) 
Don't forget, they appear only the first time we launch the application. 
Anyway, I accept both suggestions.

**Reply by LeHerosInconnu** [reply]
@Branikolog.

> > Please, do not force the player to click on a dozen windows when the game opens.
> 
> There will be just two.)))

One, two, three, four, come on everybody... :dancers: 

> Don't forget, they appear only the first time we launch the application.
> Anyway, I accept both suggestions.

Ah. :)

**Comment by Branikolog** [comment]
We can also combine our both suggestions, and make a link in system options menu to current resolution window with proposed switch in the bottom.

**Reply by LeHerosInconnu** [reply]
@Branikolog.

> We can also combine our both suggestions, and make a link in system options menu to current resolution window with proposed switch in the bottom.

I don't think that's necessary.
As you said, no one will go to this menu while the "F4" key is directly accessible.

**Comment by ihhub** [comment]
As a conclusion for this release we'll add an extra sentence to "Please remember" dialog for now.
